### PR TITLE
Allow creating wheels from source packaging.

### PR DESCRIPTION
**Pull request by arjundc-db**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?
Improve the wheel addition process for libraries that only release source packages. This allows the wheel build to use these packages and create wheels from them by overriding the default option.
(Please fill in changes proposed in this fix)

## How is this patch tested?
Manually

Following testing was done manually:
1. Create a dev wheel using this PR.
2. Uploaded to a cluster.
3. Tried downloading wheels of a simple sklearn model -> **Successful**.
4. Created a model with requirements that are known to fail (horovod, pyspark, etc).
5. Tried downloading the wheels for the new model using default behaviour.
6. The downloading fails as expected.
7. Used the following steps to set env var
`import os
os.environ["MLFLOW_WHEELED_MODEL_PIP_DOWNLOAD_OPTIONS"] = "--prefer-binary"`
9. Tried downloading again -> **Successful**

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by sueann — COMMENTED** [review]
it seems like `--prefer-binary` may not be supported by all pip versions - we may want to mention that in the docs?

is the only feasible testing for this basically manual? or are there existing tests for it already?

**Comment by WeichenXu123** [comment]
What about adding a environment variable to control this ?

By default we use `--only-binary=:all:` option, but we can allow user to set other options, like:

```
        pip_wheel_options = os.environ.get("MLFLOW_MODEL_PIP_WHEEL_OPTIONS", "--only-binary=:all:")
        download_command = (
            f"{sys.executable} -m pip wheel {pip_wheel_options} --wheel-dir={dst_path} -r"
            f"{pip_requirements_path} --no-cache-dir"
        )
```

**Review by sueann — APPROVED** [review]
the env var method sgtm - let's just make sure to add it to the docs!

**Comment by arjundc-db** [comment]
@WeichenXu123 Great suggestion: I added the env variable. Can you PTAL! I would like to merge this before the mlflow cut on monday.
@sueann 
Following testing was done manually:
1. Create a dev wheel using this PR.
2. Uploaded to a cluster.
3. Tried downloading wheels of a simple sklearn model -> **Successful**.
4. Created a model with requirements that are known to fail (horovod, pyspark, etc).
5. Tried downloading the wheels for the new model using default behaviour.
6. The downloading fails as expected.
7. Used the following steps to set env var
`import os
os.environ["MLFLOW_WHEELED_MODEL_PIP_DOWNLOAD_OPTIONS"] = "--prefer-binary"`
9. Tried downloading again -> **Successful**
I can share the model notebook if needed.

File: mlflow/models/wheeled_model.py [hunk 1]
```diff
@@ -25,6 +25,7 @@
 _WHEELS_FOLDER_NAME = "wheels"
 _ORIGINAL_REQ_FILE_NAME = "original_requirements.txt"
 _PLATFORM = "platform"
+_MLFLOW_WHEELED_MODEL_PIP_DOWNLOAD_OPTIONS = "MLFLOW_WHEELED_MODEL_PIP_DOWNLOAD_OPTIONS"
```
Comment (WeichenXu123) [inline mlflow/models/wheeled_model.py:28]: Could you define this here ? 
https://github.com/mlflow/mlflow/blob/master/mlflow/environment_variables.py and add doc for the defined `_EnvironmentVariable`


like:
```
# Add docs here ....
MLFLOW_WHEELED_MODEL_PIP_DOWNLOAD_OPTIONS = _EnvironmentVariable("MLFLOW_WHEELED_MODEL_PIP_DOWNLOAD_OPTIONS", str, "--only-binary=:all:")
```
Then we can get the env variable value by API `MLFLOW_WHEELED_MODEL_PIP_DOWNLOAD_OPTIONS.get()`

**Review by WeichenXu123 — APPROVED** [review]
LGTM once last comment is addressed.

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Pull request merged**